src/datumaro/experimental/errors.py

```diff
@@ -0,0 +1,33 @@
+# Copyright (C) 2025 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+"""
+Custom exceptions for the experimental datumaro module.
+"""
+
+
+class ArrayStructureError(TypeError):
+    """Raised when a numpy array has an improper structure for conversion.
+
+    This error indicates that the array's dtype=object structure is incorrect,
+    typically because inner arrays also have object dtype instead of a numeric dtype.
+    """
+
+    _GUIDANCE = (
+        "This typically means the numpy array structure is incorrect.\n"
+        "For fields expecting arrays (e.g., polygon_field), ensure:\n"
+        "  - The outer array has dtype=object and contains numpy arrays as elements\n"
+        "  - Each inner array has a numeric dtype (e.g., float32), NOT dtype=object\n\n"
+        "Example:\n"
+        "  # Correct: inner arrays have float32 dtype\n"
+        "  inner = np.array([[x1, y1], [x2, y2]], dtype=np.float32)\n"
+        "  outer = np.array([inner], dtype=object)\n\n"
```

Comment:
This example is wrong too, the inner array becomes `object` when you construct the outer one.
<img width="462" height="116" alt="image" src="https://github.com/user-attachments/assets/eacda6a0-e0f6-43f0-9af7-c61cd426799d" />